### DISCUSSION TITLE
osu!mania mod: Reorder

### DISCUSSION THREAD

**Discussion by aryoadhi** [ppy/osu#23505] [Ideas]
Description:

Reorder, or I would say RO (temporary name), is a mod that allows you to reorder the notes position based on lanes where the notes at. To put it simple, it's Random (RD) except you can choose the output.

This mod idea is inspired from Stepmania editor, where you can "Arbitary Shuffle" the selected notes to whichever lane position you want.

I'm not sure how do I work with public Github, and I can't compile it on my laptop due to poor performance of it. It would be great if someone can translate my idea to code.

UI idea:

![image](https://github.com/ppy/osu/assets/5029654/9e8795d5-733c-46f5-8112-c9fa683e2e05)
The idea is simple; movable blocks. The sketch is a bit crude, but you get my idea. User can click and drag the blocks to whichever order they want, so they can achieve the beatmap to be ordered as they want.

Use case:

Players can rearrange the lane orders to whichever they like. They can use it to shuffle the lanes so the harder patterns can be executed easier by simply moving the entire lane(s) to the other side. This also able to effectively emulate the mirror horizontal/vertical mods and rotation mods should they play index maps (e.g only mirroring up and down, left and right, physically turning the orientation of map by 90 degrees).

Technical:

- This mod shouldn't be compatible with existing Random mod. It should be treated as mutually exclusive.
- When using it with convert maps, the converted map should be cached first before the note reorder applies.
- When switching maps with different keycount, the note order should be reverted to default.
- osu! should detect beatmaps with special (S) style marked. If there is any, the special lane always be marked with S, before going to other keys (1, 2, 3, ...)

**Comment by roxandtol** [comment]
Perhaps in the random mod, instead of choosing a random seed / setting some random numbers and hoping you get what you want, you could just type the order you want (for example, 4321 (4k mirror), 1743256 (specific 7k random)) and vice versa
